tests/y-weak-links.tests.js

```diff
@@ -0,0 +1,475 @@
+  map1.set('link-key', link1)
+  const link2 =  /** @type {Y.WeakLink<String>} */ (map1.link('link-key'))
+  map2.set('link-link', link2)
+  const link3 =  /** @type {Y.WeakLink<String>} */ (map2.link('link-link'))
+  map3.set('link-link-link', link3)
+
+  /**
+   * @type {Array<any>}
+   */
+  let events = []
+  link3.observeDeep((e) => events = e)
+  map2.set('key', 'value2')
+  const values = events.map((e) => e.target.deref())
+  t.compare(values, ['value2'])
+}
+
+/**
+ * @param {t.TestCase} tc
+ */
+export const testDeepObserveMap = tc => {
+  // test observers in a face of linked chains of values
+  const doc = new Y.Doc()
+  /*
+     Structure:
+       - map (observed):
+         - link:<----+
+       - array:      |
+          0: nested:-+
+            - key: value
+   */
+  const map = doc.getMap('map')
+  const array = doc.getArray('array')
+
+  /**
+   * @type {Array<any>}
+   */
+  let events = []
+  map.observeDeep((e) => events = e)
+
+  const nested = new Y.Map()
+  array.insert(0, [nested])
+  const link = array.link(0)
+  map.set('link', link)
+
+  // update entry in linked map
+  events = []
+  nested.set('key', 'value')
+  t.compare(events.length, 1)
+  t.compare(events[0].target, nested)
+  t.compare(events[0].keys, new Map([['key', {action:'add', oldValue: undefined}]]))
+
+  // delete entry in linked map
+  events = []
+  nested.delete('key')
+  t.compare(events.length, 1)
+  t.compare(events[0].target, nested)
+  t.compare(events[0].keys, new Map([['key', {action:'delete', oldValue: undefined}]]))
+  
+  // delete linked map
+  array.delete(0)
+  t.compare(events.length, 1)
+  t.compare(events[0].target, link)
+}
+
+/**
+ * @param {t.TestCase} tc
+ */
+export const testDeepObserveArray = tc => {
+  // test observers in a face of linked chains of values
+  const doc = new Y.Doc()
+  /*
+     Structure:
+       - map:
+         - nested: --------+
+           - key: value    |
+       - array (observed): |
+         0: <--------------+
+   */
+  const map = doc.getMap('map')
+  const array = doc.getArray('array')
+
+  /**
+   * @type {Array<any>}
+   */
+  let events = []
+  array.observeDeep((e) => events = e)
+
+  const nested = new Y.Map()
+  map.set('nested', nested)
+  const link = map.link('nested')
+  array.insert(0, [link])
+
+  // update entry in linked map
+  events = []
+  nested.set('key', 'value')
+  t.compare(events.length, 1)
+  t.compare(events[0].target, nested)
+  t.compare(events[0].keys, new Map([['key', {action:'add', oldValue: undefined}]]))
+
+  nested.set('key', 'value2')
+  t.compare(events.length, 1)
+  t.compare(events[0].target, nested)
+  t.compare(events[0].keys, new Map([['key', {action:'update', oldValue: undefined}]]))
+
+  // delete entry in linked map
+  nested.delete('key')
+  t.compare(events.length, 1)
+  t.compare(events[0].target, nested)
+  t.compare(events[0].keys, new Map([['key', {action:'delete', oldValue: undefined}]]))
```

Comment:
This assertion fails - moreover it'll also fail in following example (deep observe over nested maps):

```js
const doc = new Y.Doc()
const outer = doc.getMap('outer')
let events = []
outer.observeDeep((e) => events = e)

const inner = new Y.Map()
outer.set('inner', inner) // clock: 1

inner.set('key', 'value1') // clock: 2
inner.set('key', 'value2') // clock: 3
inner.delete('key')
  
t.compare(events[0].keys, new Map([['key', {action:'delete', oldValue: undefined}]]))
```

I think it's a bug. The reason for that is that `event.keys` property is computed by checking if corresponding block has been added/deleted within the scope of the transaction. 

However what happens here is that we subsequently set "key" to "value1" (item clock: 1, length: 1), then "value2" (clock: 2, length: 1). When we delete "key", these two blocks get squashed and the transaction's delete set can no longer recognise squashed block as deleted (as only fragment of that block was deleted in current transaction), which in result causes `event.keys` to skip over the deleted key.

Reply:
You are gonna love this.

`event.change` is a computed property that calculates the values dynamically. So if you call it outside of the event handler (i.e. after the transaction cleanup) you're gonna have a bad time.

If you call `events[0].keys` in the observer, the value is properly cached.

I will add an error message when you do that.